src/modules/resource/monitor.c

```diff
@@ -280,6 +285,84 @@ static int monitor_reduce_self (struct monitor *monitor, bool up)
     return rc;
 }
 
+static bool notify_one_waiter (flux_t *h, int count, const flux_msg_t *msg)
+{
+    int up;
+
+    if (flux_request_unpack (msg, NULL, "{s:i}", "up", &up) < 0) {
+        if (flux_respond_error (h, msg, errno, NULL) < 0)
+            flux_log_error (h, "error resonding to monitor-waitup request");
+        return true;
+    }
+    if (up == count) {
+        if (flux_respond (h, msg, NULL) < 0)
+            flux_log_error (h, "error resonding to monitor-waitup request");
+        return true;
+    }
+    return false;
+}
+
+static bool notify_waiters (struct monitor *monitor)
+{
+    if (monitor->waiters) {
+        const flux_msg_t *msg = NULL;
+        int count = idset_count (monitor->up);
+
+        msg = zlist_first (monitor->waiters);
+        while (msg) {
+            if (notify_one_waiter (monitor->ctx->h, count, msg)) {
+                zlist_remove (monitor->waiters, (void *)msg);
```

Comment:
i don't think `zlist_remove()` is safe while iterating `zlist`.

Reply:
I asked about this as well, note that the function returns if any item is removed from the list, so this is safe. (I missed that as well, maybe a comment would help?)

Reply:
Ahh, i see.  Yeah,  a comment would be good.